python_modules/dagster/dagster/_core/definitions/mutable_partitions_definition.py

```diff
@@ -0,0 +1,86 @@
+from datetime import datetime
+from typing import Optional, Sequence
+
+import dagster._check as check
+from dagster._annotations import experimental
+from dagster._core.instance import DagsterInstance
+
+from .partition import (
+    Partition,
+    PartitionsDefinition,
+    raise_error_on_invalid_partition_key_substring,
+)
+
+
+@experimental
+class MutablePartitionsDefinition(PartitionsDefinition):
+    """
+    A partitions definition whose partition keys can be dynamically added and removed.
+
+    This is useful for cases where the set of partitions is not known at definition time,
+    but is instead determined at runtime.
+
+    Partitions can be added and removed using the `add_partitions` and `remove_partitions` methods.
+    For example:
+
+    ```
+    foo = MutablePartitionsDefinition("foo")
+
+    @sensor(job=my_job)
+    def my_sensor(context):
+        foo.add_partitions([partition_key])
+        return my_job.run_request_for_partition(partition_key)
+    ```
+    """
+
+    def __init__(self, name: str, instance: Optional[DagsterInstance] = None):
```

Comment:
My instinct here is that, in general, you shouldn't need to instantiate a `DagsterInstance` in order to construct your `Definitions`. 
- Abstractly, the set of Definitions should be independent of the instance the user is running against.
- Concretely, constructing a `DagsterInstance` can has a lot of thorns. E.g. `DagsterInstance.get` raises an error if `DAGSTER_HOME` isn't set.

Instead, methods like `add_partitions` should accept an `instance` argument.  I know this potentially adds a lot of complexity / additional lifting if `get_partitions` needs to accept an instance argument, so not 100% sure on this, but that's where I would start.